File: leetcode/c++/0121-best-time-to-buy-and-sell-stock.cc
// https://leetcode.com/problems/best-time-to-buy-and-sell-stock/
// Easy

#include<vector>

using namespace std;

class Solution {
    int buy, sell, size, gain = 0;
public:
    Solution() {
        buy = 0;
        sell = 1;
        gain = 0;
    }
    int maxProfit(vector<int>& prices) {
        size = prices.size();
        for (;sell < size; ++sell) {
            gain = prices[sell] - prices[buy] > gain ? prices[sell] - prices[buy] : gain;
            if (prices[buy] > prices[sell]) {
                buy = sell;
            }
        }
        return gain;
    }
};

File: leetcode/python/0121-best-time-to-buy-and-sell-stock.py
# https://leetcode.com/problems/best-time-to-buy-and-sell-stock/
# Easy

# Keep track of smallest price so far, and continuously compare current price with smallest price
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        ret, smallest = 0, prices[0]
        for n in prices[1:]:
            smallest = min(smallest, n)
            ret = max(ret, n-smallest)
        return ret